File: tools/dump_geo.py
#!/usr/bin/env python3
"""把原版 Python 仿真的逐帧图元导出成 JSON，供 tools/verify.js 与 JS 版对拍。

不画图、不开窗口：用一只"黑洞"画布接住 stamp/drop，只统计几何。
每帧记录 last_geos（含临死元素的图元数）、存活元素数，以及**存活元素**
这一帧的图元列表 —— 与 JS 版 Show.frameGeos 的口径完全一致。

用法：python3 tools/dump_geo.py --scene classic --seed 7 --frames 90
"""
import argparse
import hashlib
import importlib.util
import json
import os
import random
import struct
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
REF = os.path.join(HERE, os.pardir, "reference", "turtle_fireworks.py")
DT = 1.0 / 60.0


class SinkCanvas:
    """只统计，不做任何光栅化（SoftwareCanvas 的 drop 是线性扫描，这里省掉）。"""

    def stamp(self, g):
        return None

    def drop(self, sid):
        pass


def load_reference():
    spec = importlib.util.spec_from_file_location("turtle_fireworks", REF)
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod


def rng_digest(rng):
    """MT19937 全部 624 个状态字 + 下标 的摘要：逐帧比它就能证明随机数流完全一致
    （比逐个图元比对更强 —— 任何一次 random/randint/uniform 的差异都会暴露）。"""
    state = rng.getstate()[1]
    return hashlib.sha256(
        struct.pack("<%dI" % len(state), *[v & 0xFFFFFFFF for v in state])
    ).hexdigest()[:16]


def pack(g):
    return [g.shape, g.color[0], g.color[1], g.color[2],
            g.x, g.y, g.heading, g.wid, g.leng]


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--scene", default="classic",
                    choices=("classic", "random", "original"))
    ap.add_argument("--seed", type=int, default=7)
    ap.add_argument("--frames", type=int, default=90)
    ap.add_argument("--width", type=int, default=924)
    ap.add_argument("--height", type=int, default=691)
    ap.add_argument("--max-geos", type=int, default=130)
    ap.add_argument("--out", default=None, help="输出文件（默认 stdout）")
    args = ap.parse_args()

    tf = load_reference()
    rng = random.Random(args.seed)
    show = tf.Show(SinkCanvas(), tf.BG, args.width, args.height, rng,
                   max_geos=args.max_geos)
    tf.build_show(show, args.scene)

    frames = []

    def snapshot(i):
        geos = []
        for e in show.elements:               # commit() 把本帧图元留在 e.prev 里
            for g in e.prev:
                geos.append(pack(g))
        frames.append({"i": i, "time": show.time, "lastGeos": show.last_geos,
                       "elements": len(show.elements), "rng": rng_digest(rng),
                       "geos": geos})

    show.step(0.0)                            # 先把初始几何画上（dt=0 不推进物理）
    snapshot(0)
    for i in range(1, max(1, args.frames)):
        show.step(DT)                         # 固定 60fps 步长
        snapshot(i)

    data = {"scene": args.scene, "seed": args.seed, "width": args.width,
            "height": args.height, "dt": DT, "frames": frames}
    text = json.dumps(data, separators=(",", ":"))
    if args.out:
        with open(args.out, "w") as fh:
            fh.write(text)
    else:
        sys.stdout.write(text + "\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/dump_geo.js
#!/usr/bin/env node
/* 把 JS 版仿真的逐帧图元导出成 JSON（协议与 tools/dump_geo.py 完全一致）。 */
'use strict';

const crypto = require('crypto');
const { loadSim } = require('./loader');
const FW = loadSim();

const DT = 1 / 60;

/** MT19937 全部 624 个状态字 + 下标 的摘要（与 tools/dump_geo.py 的字节布局一致）。 */
function rngDigest(rng) {
  const buf = Buffer.allocUnsafe((rng.mt.length + 1) * 4);
  for (let i = 0; i < rng.mt.length; i++) buf.writeUInt32LE(rng.mt[i], i * 4);
  buf.writeUInt32LE(rng.mti, rng.mt.length * 4);
  return crypto.createHash('sha256').update(buf).digest('hex').slice(0, 16);
}

function parseArgs(argv) {
  const a = { scene: 'classic', seed: 7, frames: 90, width: 924, height: 691,
              maxGeos: 130, out: null, dense: false };
  for (let i = 2; i < argv.length; i += 2) {
    const k = argv[i].replace(/^--/, ''), v = argv[i + 1];
    if (k === 'scene') a.scene = v;
    else if (k === 'seed') a.seed = Number(v);
    else if (k === 'frames') a.frames = Number(v);
    else if (k === 'width') a.width = Number(v);
    else if (k === 'height') a.height = Number(v);
    else if (k === 'max-geos') a.maxGeos = Number(v);
    else if (k === 'dense') a.dense = (v === '1' || v === 'true');
    else if (k === 'out') a.out = v;
  }
  return a;
}

function pack(g) {
  const c = g.color;
  return [g.shape, c[0], c[1], c[2], g.x, g.y, g.heading, g.wid, g.leng];
}

const args = parseArgs(process.argv);
const rng = new FW.rng.Random(args.seed);
const show = new FW.show.Show(args.width, args.height, rng,
                              { maxGeos: args.maxGeos, denseTrails: args.dense });
FW.show.build(show, args.scene);

const frames = [];
function snapshot(i) {
  const geos = [];
  let segs = 0;
  for (const g of show.frameGeos) {
    if (g.segs) segs += g.segs.length; else geos.push(pack(g));
  }
  frames.push({ i, time: show.time, lastGeos: show.lastGeos, denseSegs: segs,
                elements: show.elements.length, rng: rngDigest(rng), geos });
}

show.step(0.0);
snapshot(0);
for (let i = 1; i < Math.max(1, args.frames); i++) {
  show.step(DT);
  snapshot(i);
}

const data = { scene: args.scene, seed: args.seed, width: args.width,
               height: args.height, dt: DT, frames };
const text = JSON.stringify(data);
if (args.out) require('fs').writeFileSync(args.out, text);
else process.stdout.write(text + '\n');
